src/lucky/subdomain.cr

```diff
@@ -0,0 +1,71 @@
+module Lucky::Subdomain
+  # Taken from https://github.com/rails/rails/blob/afc6abb674b51717dac39ea4d9e2252d7e40d060/actionpack/lib/action_dispatch/http/url.rb#L8
+  IP_HOST_REGEXP = /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/
+
+  Habitat.create do
+    # tld_length is the number of Top Level Domain segments separated by periods
+    # the default is 1 because most domains end in ".com" or ".org"
+    # The tld_length should be changed to 2 when you have a ".co.uk" domain for example
+    # It can also be changed to 0 for local development so that you can use `tenant.localhost:3000`
+    setting tld_length : Int32 = 1
+  end
+
+  alias Matcher = String | Regex | Bool | Array(String | Regex) | Array(String) | Array(Regex)
+
+  # Sets up a subdomain requirement for an action
+  #
+  # ```
+  # register_subdomain                                    # subdomain required but can be anything
+  # register_subdomain "admin"                            # subdomain required and must equal "admin"
+  # register_subdomain /(dev|qa|prod)/                    # subdomain required and must match regex
+  # register_subdomain ["tenant1", "tenant2", /tenant\d/] # subdomain required and must match one of the items in the array
```

Comment:
I love all the different options. Thinking about how I currently use subdomains, I only need a subdomain on my routes when I'm in staging. In production, we don't use subdomains. Is that possible to do? I'm thinking when you do `qa.luckyframework.org`, that would only be on the QA environment, and not in production.. So we need to be able to say `register_subdomain "qa" if LuckyEnv.qa?` or whatever...

Reply:
I have pushed an update to better clarify. The macro is now called `require_subdomain`.

In this situation you don't actually require a subdomain, you just do have a subdomain in certain scenarios. In those situations you wouldn't use `require_subdomain` but you can still check if a subdomain was provided by calling `subdomain?`. Calling `subdomain` will raise a compile time error though.

Reply:
Cool. That gives us a nice escape hatch then too.